Mathlib/CategoryTheory/Galois/Prorepresentability.lean

```diff
@@ -0,0 +1,235 @@
+/-
+Copyright (c) 2024 Christian Merten. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christian Merten
+-/
+import Mathlib.Algebra.Category.GroupCat.Basic
+import Mathlib.CategoryTheory.Galois.Decomposition
+import Mathlib.CategoryTheory.Limits.FunctorCategory
+
+/-!
+# Pro-Representability of fiber functors
+
+We show that any fiber functor is pro-representable, i.e. there exists a pro-object
+`X : I ⥤ C` such that `F` is naturally isomorphic to `X ⋙ coyoneda`.
+
+## Main definitions
+
+- `PointedGaloisObject`: the category of pointed Galois objects
+- `PointedGaloisObject.cocone`: a cocone on `(PointedGaloisObject.incl F).op ≫ coyoneda` with
+  point `F ⋙ FintypeCat.incl`.
+- `autGaloisSystem`: the system of automorphism groups indexed by the pointed Galois objects.
+
+## Main results
+
+- `PointedGaloisObject.isColimit`: the cocone `PointedGaloisObject.cocone` is a colimit cocone.
+
+## References
+
+* [lenstraGSchemes]: H. W. Lenstra. Galois theory for schemes.
+
+-/
+
+universe u₁ u₂ w
+
+namespace CategoryTheory
+
+namespace PreGaloisCategory
+
+open Limits Functor
+
+variable {C : Type u₁} [Category.{u₂} C] [GaloisCategory C]
+variable (F : C ⥤ FintypeCat.{u₂}) [FiberFunctor F]
+
+/-- A pointed Galois object is a Galois object with a fixed point of its fiber. -/
+structure PointedGaloisObject : Type (max u₁ u₂) where
+  /-- The underlying object of `C`. -/
+  obj : C
+  /-- An element of the fiber of `obj`. -/
+  pt : F.obj obj
+  /-- `obj` is Galois. -/
+  isGalois : IsGalois obj := by infer_instance
+
+namespace PointedGaloisObject
+
+attribute [instance] isGalois
+
+instance (X : PointedGaloisObject F) : CoeDep (PointedGaloisObject F) X C where
+  coe := X.obj
+
+variable {F} in
+/-- The type of homomorphisms between two pointed Galois objects. This is a homomorphism
+of the underlying objects of `C` that maps the distinguished points to each other. -/
+structure Hom (A B : PointedGaloisObject F) where
+  /-- The underlying homomorphism of `C`. -/
+  val : A.obj ⟶ B.obj
+  /-- The distinguished point of `A` is mapped to the distinguished point of `B`. -/
+  comp : F.map val A.pt = B.pt
+
+/-- The category of pointed Galois objects. -/
+instance : Category.{u₂} (PointedGaloisObject F) where
+  Hom A B := Hom A B
+  id A := ⟨𝟙 (A : C), by simp⟩
+  comp {A B C} f g := by
+    refine ⟨f.val ≫ g.val, ?_⟩
+    simp only [F.map_comp, FintypeCat.comp_apply, f.comp, g.comp]
+
+instance {A B : PointedGaloisObject F} : Coe (Hom A B) (A.obj ⟶ B.obj) where
+  coe f := f.val
+
+variable {F}
+
+@[ext]
+lemma Hom.ext {A B : PointedGaloisObject F} {f g : A ⟶ B} (_ : f.val = g.val) : f = g :=
+  match f, g with | ⟨_, _⟩, ⟨_, _⟩ => by congr
```

Comment:
The following syntax seems easier to read:
```suggestion
  cases f; cases g; congr
```

Reply:
I later learned (from looking at one of your PRs) that I can put `@[ext]` on a structure, so I introduced a `hom_ext` instead which just uses the automatically generated one. Is that the preferred way?

Reply:
Yes, any of these two options is good.